File: ai_autopilot/routing/task_router.py
"""Classify work items and route them to Claude Code skills (ported from ``TaskRouter``)."""

from __future__ import annotations

import re

from ai_autopilot.logging_config import get_logger
from ai_autopilot.models import TaskCategory, WorkItemInfo

# Skill command templates — these invoke skills already in .claude/skills/.
_SKILL_MAP: dict[TaskCategory, str] = {
    TaskCategory.BACKEND_TASK: "/crud-full-stack {id}",
    TaskCategory.FRONTEND_TASK: "/fe-module {id}",
    TaskCategory.DATABASE_TASK: "/sql-migration {id}",
    TaskCategory.BUG: "/bugfix-workflow {id}",
    TaskCategory.TEST_TASK: "/qc-test-management {id}",
    TaskCategory.REQUIREMENT: "/analyze-requirement {id}",
}

_BACKEND_KW = ("API", "CONTROLLER", "SERVICE", "ENTITY", "MIGRATION", "ENDPOINT")
_FRONTEND_KW = ("COMPONENT", "PAGE", "FORM", "UI", "ANGULAR", "SCREEN")


class TaskRouter:
    def __init__(self) -> None:
        self._log = get_logger("routing.task_router")

    def classify(self, item: WorkItemInfo) -> WorkItemInfo:
        item.category = self._detect_category(item)
        return item

    def route(self, item: WorkItemInfo) -> str | None:
        """Map a classified item to a skill command, or None if unroutable."""
        if item.category is TaskCategory.UNKNOWN:
            self._log.warning("cannot route", id=item.id, category=str(item.category))
            return None
        template = _SKILL_MAP.get(item.category)
        if template is None:
            return None
        return template.replace("{id}", str(item.id))

    @staticmethod
    def _detect_category(item: WorkItemInfo) -> TaskCategory:
        title = item.title.upper()
        tags = " ".join(item.tags).upper()
        item_type = item.work_item_type.upper()

        if item_type == "BUG":
            return TaskCategory.BUG

        if _has_prefix(title, "[BE]") or _has_tag(tags, "BE") or _has_tag(tags, "BACKEND"):
            return TaskCategory.BACKEND_TASK
        if _has_prefix(title, "[FE]") or _has_tag(tags, "FE") or _has_tag(tags, "FRONTEND"):
            return TaskCategory.FRONTEND_TASK
        if _has_prefix(title, "[DB]") or _has_tag(tags, "DB") or _has_tag(tags, "DATABASE"):
            return TaskCategory.DATABASE_TASK
        if (
            _has_prefix(title, "[QC]")
            or _has_prefix(title, "[TEST]")
            or _has_tag(tags, "QC")
            or _has_tag(tags, "TEST")
        ):
            return TaskCategory.TEST_TASK

        if item_type in ("USER STORY", "REQUIREMENT", "FEATURE"):
            return TaskCategory.REQUIREMENT

        if item_type == "TASK":
            if any(k in title for k in _BACKEND_KW):
                return TaskCategory.BACKEND_TASK
            if any(k in title for k in _FRONTEND_KW):
                return TaskCategory.FRONTEND_TASK

        return TaskCategory.UNKNOWN


def _has_prefix(title: str, prefix: str) -> bool:
    return title.startswith(prefix)


def _has_tag(tags: str, tag: str) -> bool:
    return re.search(rf"\b{re.escape(tag)}\b", tags) is not None

File: legacy-dotnet/Routing/TaskRouter.cs
using System.Text.RegularExpressions;
using AdoAutopilot.Models;

namespace AdoAutopilot.Routing;

/// <summary>
/// Classifies work items and routes them to the appropriate Claude Code skill.
/// </summary>
public class TaskRouter
{
    private readonly ILogger<TaskRouter> _logger;

    // Skill command templates — these invoke skills already in .claude/skills/
    private static readonly Dictionary<TaskCategory, string> SkillMap = new()
    {
        [TaskCategory.BackendTask]  = "/implement-task-be {id}",
        [TaskCategory.FrontendTask] = "/implement-task-fe {id}",
        [TaskCategory.Bug]          = "/bugfix-workflow {id}",
        [TaskCategory.TestTask]     = "/qc-test-management {id}",
        [TaskCategory.Requirement]  = "/analyze-requirement {id}",
    };

    public TaskRouter(ILogger<TaskRouter> logger)
    {
        _logger = logger;
    }

    /// <summary>
    /// Classify a work item into a TaskCategory based on title, type, and tags.
    /// </summary>
    public WorkItemInfo Classify(WorkItemInfo item)
    {
        item.Category = DetectCategory(item);
        return item;
    }

    /// <summary>
    /// Route a classified work item to a Claude skill command string.
    /// Returns null if no skill matches.
    /// </summary>
    public string? Route(WorkItemInfo item)
    {
        if (item.Category == TaskCategory.Unknown || item.Category == TaskCategory.DatabaseTask)
        {
            _logger.LogWarning("Cannot route #{Id}: category={Category}", item.Id, item.Category);
            return null;
        }

        if (!SkillMap.TryGetValue(item.Category, out var template))
            return null;

        return template.Replace("{id}", item.Id.ToString());
    }

    private static TaskCategory DetectCategory(WorkItemInfo item)
    {
        var title = item.Title.ToUpperInvariant();
        var tags = string.Join(" ", item.Tags).ToUpperInvariant();
        var type = item.WorkItemType.ToUpperInvariant();

        // Bug type takes priority
        if (type == "BUG") return TaskCategory.Bug;

        // Check by title prefix — NOIS convention: [BE], [FE], [DB], [QC]
        if (HasPrefix(title, "[BE]") || HasTag(tags, "BE") || HasTag(tags, "BACKEND"))
            return TaskCategory.BackendTask;

        if (HasPrefix(title, "[FE]") || HasTag(tags, "FE") || HasTag(tags, "FRONTEND"))
            return TaskCategory.FrontendTask;

        if (HasPrefix(title, "[DB]") || HasTag(tags, "DB") || HasTag(tags, "DATABASE"))
            return TaskCategory.DatabaseTask;

        if (HasPrefix(title, "[QC]") || HasPrefix(title, "[TEST]") || HasTag(tags, "QC") || HasTag(tags, "TEST"))
            return TaskCategory.TestTask;

        // Check work item type
        if (type is "USER STORY" or "REQUIREMENT" or "FEATURE")
            return TaskCategory.Requirement;

        if (type == "TASK")
        {
            // Try to infer from title keywords
            if (ContainsAny(title, "API", "CONTROLLER", "SERVICE", "ENTITY", "MIGRATION", "ENDPOINT"))
                return TaskCategory.BackendTask;
            if (ContainsAny(title, "COMPONENT", "PAGE", "FORM", "UI", "ANGULAR", "SCREEN"))
                return TaskCategory.FrontendTask;
        }

        return TaskCategory.Unknown;
    }

    private static bool HasPrefix(string title, string prefix)
        => title.StartsWith(prefix, StringComparison.OrdinalIgnoreCase);

    private static bool HasTag(string tags, string tag)
        => Regex.IsMatch(tags, $@"\b{Regex.Escape(tag)}\b");

    private static bool ContainsAny(string text, params string[] keywords)
        => keywords.Any(k => text.Contains(k, StringComparison.OrdinalIgnoreCase));
}
